gnocchi/storage/ceph.py

```diff
@@ -173,13 +176,25 @@ def _list_split_keys(self, metric, aggregation, granularity, version=3):
             except rados.ObjectNotFound:
                 raise storage.MetricDoesNotExist(metric)
 
-            keys = set()
-            granularity = str(utils.timespan_total_seconds(granularity))
-            for name, value in omaps:
-                meta = name.split('_')
-                if (aggregation == meta[3] and granularity == meta[4]
-                        and self._version_check(name, version)):
-                    keys.add(meta[2])
+            raw_keys = [name.split("_")
+                        for name, value in omaps
+                        if self._version_check(name, version)]
+            keys = collections.defaultdict(set)
+            if not raw_keys:
+                return keys
+            zipped = list(zip(*raw_keys))
+            k_timestamps = utils.to_timestamps(zipped[2])
+            k_methods = zipped[3]
+            k_granularities = list(map(utils.to_timespan, zipped[4]))
+
+            for timestamp, method, granularity in six.moves.zip(
+                    k_timestamps, k_methods, k_granularities):
+                for aggregation in aggregations:
```

Comment:
this seems simpler to just keep original code, and just build a set of (granularity, method) pairs to represent Agg, then you can just check if omap is in the set rather than all these loops?

Reply:
I don't want to abuse, but i'm not sure I get it; would you mind writing the code snippet you have in mind? :)

Reply:
like this: https://gist.github.com/chungg/46bb333e6a1f9dc862d9b799ffe6a874

disclaimer: i never tested.

Reply:
oh, you moved the numpy broadcasting into here...

Reply:
This would not work as the timestamp needs to be parsed. Your code does not.

My code here converts all the timestamps in one single call (L186). I did that to make it fast because I knew otherwise you would (rightfully) complain. Yes I always think of you when I code. :)

Reply:
we can still convert this to "if x in y" rather than a for loop no? or at the very least, add a break once we fined match all the additional loops concern me but it might be only way as you said.

Reply:
We list and parse everything, and then iterate over every key and check if it's interesting for any aggregation.

We can write this code different with a `if x in y` but this is just syntatic sugar, as it'll iterate in `y` anyway. There's no magic here.

Reply:
that's fair regarding syntax but we can still break once we find match as i don't see how there can be multiple matches

Reply:
Oh you're right I think. I missed that.